p2p: track AddrMan totals by network and table, improve precision of adding fixed seeds

mzumsande opened the pull request:
AddrMan currently doesn't track the number of its entries by network, it only knows the total number of addresses. This PR makes AddrMan keep track of these numbers, which would be helpful for multiple things:

1. Allow to specifically add fixed seeds to AddrMan of networks where we don't have any addresses yet - even if AddrMan as a whole is not empty (partly fixing #26035). This is in particular helpful if the user abruptly changes `-onlynet` settings (such that addrs that used to be reachable are no longer and vice versa), in which case they currently could get stuck and not find any outbound peers. The second commit of this PR implements this.
1. (Future work): Add logic for automatic connection management with respect to networks - such as making attempts to have at least one connection to each reachable network as suggested [here](https://github.com/bitcoin/bitcoin/issues/26035#issuecomment-1249420209). This would involve requesting an address from a particular network from AddrMan, and expanding its corresponding function `AddrMan::Select()`  to do this requires internal knowledge of the current number of addresses for each network and table to avoid getting stuck in endless loops.
1. (Future work): Perhaps display the totals to users. At least I would find this helpful to debug, the existing option (`./bitcoin-cli -addrinfo`) is rather indirect by doing the aggregation itself in each call, doesn't distinguish between new and tried, and being based on `AddrMan::GetAddr()` it's also subject to a quality filter which we probably don't want in this spot.

mzumsande committed:
addrman: add function to return size by network and table

For now, the new functionality will be used in the context of
querying fixed seeds. Other possible applications for
future changes is the use in the context of making automatic
connections to specific networks, or making more detailed info
about addrman accessible via rpc.

mzumsande committed:
net: Load fixed seeds from reachable networks for which we don't have addresses

Previously, we'd only load fixed seeds if we'd not
know any addresses at all. This change makes it possible
to change -onlynet abruptly, e.g. from -onlynet=onion to
-onlynet=i2p and still find peers.

amitiuttarwar commented:
approach ACK. relatively simple patch to ameliorate the issue of address discovery when switching networks

naumenkogs commented on src/addrman.cpp:
```diff
@@ -1065,6 +1099,15 @@ int AddrManImpl::CheckAddrman() const
+        cnt_tried += net.second.second;
+    }
+    if (cnt_new != nNew || cnt_tried != nTried) {
+        return -20;
```
Perhaps it's a good opportunity to start using a named variable? :)

mzumsande replied:
Do you mean getting rid of the return codes and return something like error strings instead? 
Hmm, not sure if this wouldn't better be a separate PR.

mzumsande replied:
ok - I think refactoring the codes might make sense, but if so I'd prefer to do this everywhere at once (and in a separate refactoring PR)

naumenkogs commented on src/addrman.cpp:
```diff
@@ -962,6 +970,32 @@ std::optional<AddressPosition> AddrManImpl::FindAddressEntry_(const CAddress& ad
+        auto entry = it->second;
+        if (!in_new.has_value()) {
+            return entry.first + entry.second;
+        } else if (in_new == true) {
```
Here and above, maybe use `return (in_new == true) ? entry.first : entry.second;`?

mzumsande replied:
Changed it to the ternary operator.

naumenkogs commented on src/addrman.cpp:
```diff
@@ -1065,6 +1099,15 @@ int AddrManImpl::CheckAddrman() const
+        cnt_new += net.second.first;
+        cnt_tried += net.second.second;
+    }
+    if (cnt_new != nNew || cnt_tried != nTried) {
```
Given this dependency, should we drop `nNew` and `nTried` altogether?

mzumsande replied:
I think that's not so streighforward - `nNew` and `nTried` are part of the serialization after all. Also, I don't like that then we'd need to build the sum over all networks every time we needed `nNew` or `nTried` (as the check code does).

naumenkogs replied:
>Also, I don't like that then we'd need to build the sum over all networks every time we needed nNew or nTried (as the check code does).

Well, I don't think something being inside the `Check` is a good justification — it's the other way around, `Check` is needed to make sure those values are sane. Furthermore, the sum over all networks is 5 addition operations, right? 

However, I agree that dropping these values might be not trivial due to serialization. Not sure I'll find time to prototype that, but maybe someone will :) Otherwise feel free to close this suggestion.

naumenkogs replied:
Also, see [this](https://github.com/bitcoin/bitcoin/pull/26847#discussion_r1065149504) :)

mzumsande replied:
I've changed the check such that it no longer compares with `nNew` and `nTried` but does a recount, see below for more context (of course the relation is still there...).

naumenkogs commented on src/net.h:
```diff
@@ -969,6 +969,12 @@ class CConnman
+     Return networks for which we have no addresses in addrman and therefore
+     may require loading fixed seeds.
+     */
+    std::vector<Network> GetFixedSeedNetworks() const;
```
I wish `for which we have no addresses` was captured by the function name somehow. Perhaps `GetRelevantFixedSeedNetworks`?

mzumsande replied:
Renamed it, pretty sure I even used that name in an earlier iteration.

brunoerg reviewed (commented):
Concept ACK

mzumsande commented on src/addrman.cpp:
```diff
@@ -1065,6 +1095,15 @@ int AddrManImpl::CheckAddrman() const
     if (nKey.IsNull())
         return -16;
 
+    int cnt_new{0};
```
That would get me a compiler warning `comparison of integer expressions of different signedness`, since `nNew` and `nTried` are `int`. But I'll change the sanity check to compare with the actual totals instead of `nNew` and `nTried`, as suggested by you below.

mzumsande commented on src/addrman_impl.h:
```diff
@@ -215,6 +217,9 @@ class AddrManImpl
     const NetGroupManager& m_netgroupman;
 
+    /** Number of entries in addrman per network and new/tried table. */
+    std::unordered_map<Network, std::pair<size_t, size_t>> map_counts GUARDED_BY(cs);
```
renamed

mzumsande commented on src/addrman.cpp:
```diff
@@ -962,6 +970,28 @@ std::optional<AddressPosition> AddrManImpl::FindAddressEntry_(const CAddress& ad
+        if (!in_new.has_value()) {
+            return nNew + nTried;
+        } else {
+            return (in_new == true) ? nNew : nTried;
```
done

mzumsande commented on src/addrman.cpp:
```diff
@@ -1065,6 +1095,15 @@ int AddrManImpl::CheckAddrman() const
 
+    int cnt_new{0};
+    int cnt_tried{0};
+    for (auto net : map_counts) {
```
I changed this to actually recompute the totals - while doing so I noticed that it is possible that it's possible that there are "empty" entries  (0 new and 0 tried) in `m_map_counts` if an addr of some network was first added and later removed again from. That's not a problem in my opinion, but it means that the size of the map is not necessarily reliable. The changed check should work with this, potentially also adding an entry empty to `local_counts`.

naumenkogs commented on src/addrman.h:
```diff
@@ -102,6 +102,13 @@ class AddrMan
     size_t size() const;
 
+    /**
+    * Return the number of (unique) addresses, optionally filtered by network
```
What do you mean by unique here? How they could not be unique?

mzumsande commented on src/addrman.cpp:
```diff
@@ -1065,6 +1098,17 @@ int AddrManImpl::CheckAddrman() const
+        return -20;
+    }
+    for (const auto& [net, count] : m_map_counts) {
+        if (local_counts[net].first != count.first || local_counts[net].second != count.second) {
```
Yes, It definitely shouldn't be possible and would mean that there is a serious bug in addrman, but I think that within the `Check()` code, we definitely want to abort. After all, `Check()`  doesn't normally run in production (`DEFAULT_ADDRMAN_CONSISTENCY_CHECKS == 0`) and needs to be activated by the user.

vasild commented on src/net.cpp:
```diff
@@ -1650,14 +1664,12 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 seed_addrs.erase(std::remove_if(seed_addrs.begin(), seed_addrs.end(),
-                                               [](const CAddress& addr) { return !IsReachable(addr); }),
-                                seed_addrs.end());
+                                                [fixed_seed_networks](const CAddress& addr) { return std::find(fixed_seed_networks.begin(), fixed_seed_networks.end(), addr.GetNetwork()) == fixed_seed_networks.end(); }),
```
A vector is usually not cheap to copy. Better capture it by reference:
```suggestion
                                                [&fixed_seed_networks](const CAddress& addr) { return std::find(fixed_seed_networks.begin(), fixed_seed_networks.end(), addr.GetNetwork()) == fixed_seed_networks.end(); }),
```

vasild commented on src/net.cpp:
```diff
@@ -1650,14 +1664,12 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 seed_addrs.erase(std::remove_if(seed_addrs.begin(), seed_addrs.end(),
-                                               [](const CAddress& addr) { return !IsReachable(addr); }),
-                                seed_addrs.end());
+                                                [fixed_seed_networks](const CAddress& addr) { return std::find(fixed_seed_networks.begin(), fixed_seed_networks.end(), addr.GetNetwork()) == fixed_seed_networks.end(); }),
```
This will iterate `fixed_seed_networks` for every seed address: `O(num_addresses * num_networks)`. I know the list of networks is tiny and this can probably be ignored wrt performance, but it can easily be made `O(num_addresses)`. Also, if `std::unordered_set` is used, the code will be shorter:

```cpp
std::find(fixed_seed_networks.begin(), fixed_seed_networks.end(), addr.GetNetwork()) == fixed_seed_networks.end();
```

would become:

```cpp
fixed_seed_networks.count(addr.GetNetwork()) == 0;
```

mzumsande replied:
changed it to unordered_set as suggested, thanks!

vasild commented on src/net.cpp:
```diff
@@ -1575,6 +1575,19 @@ int CConnman::GetExtraBlockRelayCount() const
     return std::max(block_relay_peers - m_max_outbound_block_relay, 0);
 }
 
+std::vector<Network> CConnman::GetRelevantFixedSeedNetworks() const
+{
+    std::vector<Network> networks{};
+    for (int n = 0; n < NET_MAX; n++) {
+        enum Network net = (enum Network)n;
+        if (net == NET_UNROUTABLE || net == NET_INTERNAL) continue;
+        if (IsReachable(net) && addrman.size(net, std::nullopt) == 0) {
+            networks.push_back(net);
+        }
+    }
+    return networks;
+}
```
This does not need to be a method of `CConnman` because it does not access any of its members (other than `addrman`). Would it be more appropriate to have it as `AddrMan` method? Then the public interface of `AddrMan` need not be changed by this PR.

mzumsande replied:
I'm not sure if this belongs into AddrMan because of the `IsReachable` - I think the information which networks are reachable belongs into `net`, and possibly at some later point into `CConnman` (I know `vfLimited` is currently a global, but this might change). I don't think it makes sense that `AddrMan` should know or deal with this information, so I think I prefer leaving this out of `AddrMan` and reducing the change of the public interface by combining the `size` functions as suggested by Amiti below.

vasild replied:
What about a standalone function?

mzumsande replied:
I'd be ok with that, but then we'd have to pass it addrman, which is a member of `CConnman`, so I'm not sure why that is better.

vasild reviewed (commented):
Concept ACK

amitiuttarwar committed:
addrman, refactor: combine two size functions

The functionality of the old size() is covered by the new Size()
when no arguments are specified, so this does not change behavior.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar commented on src/net.h:
```diff
@@ -969,6 +969,12 @@ class CConnman
+     Return networks for which we have no addresses in addrman and therefore
+     may require loading fixed seeds.
+     */
+    std::vector<Network> GetRelevantFixedSeedNetworks() const;
```
naming: "fixed seeds" seems odd to put in the function name because that's the caller - doesn't really have to do with the logic of the function. I think the functionality would better be captured with something like `GetReachableEmptyNetworks()`

mzumsande replied:
renamed, thanks

amitiuttarwar commented:
code looks good. I went through all the addrman touch points to the new & tried tables to double check nothing was missed. tried my best to think through edge cases of what could go wrong and don't have anything :) 

left a bunch of review comments that are all style related, no blockers. 

another style related suggestion: what do you think about consolidating the two `size()` functions? adding default params to the newly introduced function means you can remove the old one & none* of the call sites need to change. I tried it out [here](https://github.com/amitiuttarwar/bitcoin/commit/ab188a62bd18668f5a9174db4dadfc398bc02b76). 
[*] none of the bitcoind callers. one test call site failed because of the shift from `vRandom.size` to `nNew + nTried`. the test is checking a corrupted addrman, and I don't think the size is testing anything meaningful there, so I removed that check.

vasild commented on src/addrman.cpp:
```diff
@@ -986,6 +1016,7 @@ int AddrManImpl::CheckAddrman() const
 
     std::unordered_set<int> setTried;
     std::unordered_map<int, int> mapNew;
+    std::unordered_map<Network, std::pair<size_t, size_t>> local_counts;
```
`m_map_counts` too. The less first/second the better.

mzumsande replied:
I like the suggestion and changed it to use the struct in both spots.

vasild commented:
> ... of what could go wrong and don't have anything ...

Oh, well.

> _When nothing can possibly go wrong, it will._

:)

I guess `nNew` and `m_map_count[].first` could go out of sync in the future. It is good that now they are always updated next to each other (or within 1-2 lines). Further enforcing of this would be to add a new method that modifies both and only change them through this method. E.g.

```cpp
void IncNew(ssize_t delta, Network net)
{
    nNew += delta;
    m_map_count[net].first += delta;
}
```

Or ditch `nNew` and sum `m_map_count[].first` whenever we would read `nNew`, via a new method like `size_t CountNew() const`? Performance considerations?

amitiuttarwar commented:
ah, just to clarify, I meant I couldn't identify anything that could go wrong in the _current_ version of the code. ofc things can always go awry in the future, which is why code style is useful to help reduce the chances of that happening :) 

> I guess nNew and m_map_count[].first could go out of sync in the future. It is good that now they are always updated next to each other (or within 1-2 lines). Further enforcing of this would be to add a new method that modifies both and only change them through this method.

agreed that the counts of `nNew` (and also `nTried`) could go out of sync with the totals in `m_map_count`. it can also be problematic if these values are in sync with each other, but out of sync with the actual values stored on the new / tried tables & other addrman structures. 

the code added to `CheckAddrman` are helpful because they offer a programatic testing against this inadvertently changing in the future. I don't think having specific "accessor" methods makes the most sense because we already have functions that offer logical coupling of creating & removing addrs on addrman*. also, it would require 4 additional functions (increment & decrement / new & tried).

[*] for eg. other than unserialization, all additions to the new table go through `Addrman::Add`, which calls `AddSingle` which uses `Create` as a helper to construct the `AddrInfo` object & start persisting to internal data structs. my comment [here](https://github.com/bitcoin/bitcoin/pull/26847#discussion_r1070146333) suggested moving incrementing both counts inside the `Create` function. 

> Or ditch nNew and sum m_map_count[].first whenever we would read nNew, via a new method like size_t CountNew() const? Performance considerations?

this one I'm more open to. martin pointed out some difficulties in that approach [here](https://github.com/bitcoin/bitcoin/pull/26847#discussion_r1065041916), and I personally think its nbd either way.

mzumsande commented on src/addrman_impl.h:
```diff
@@ -215,6 +217,9 @@ class AddrManImpl
     const NetGroupManager& m_netgroupman;
 
+    /** Number of entries in addrman per network and new/tried table. */
+    std::unordered_map<Network, std::pair<size_t, size_t>> m_map_counts GUARDED_BY(cs);
```
done

mzumsande commented on src/addrman.cpp:
```diff
@@ -1074,6 +1118,15 @@ size_t AddrManImpl::size() const
+size_t AddrManImpl::size(std::optional<Network> net, std::optional<bool> in_new) const
+{
+    LOCK(cs);
+    Check();
+    auto ret = size_(net, in_new);
+    Check();
```
It's not needed, but this pattern is also used for other const methods (`GetAddr()`, `Select()`), plus we have a few mutables in AddrMan - since `Check()` is only run if the user enables it, I guess it's better to invoke it once too often rather than not?

mzumsande committed:
addrman, refactor: move count increment into Create()

Create() is only called in one spot, so this doesn't
change behavior.

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande commented on src/addrman.h:
```diff
@@ -102,6 +102,13 @@ class AddrMan
     //! Return the number of (unique) addresses in all tables.
     size_t size() const;
 
+    /**
+    * Return the number of (unique) addresses, optionally filtered by network
+    * and/or addrman table (new if true, tried if false).
+    * If any of the optional args are nullopt, don't filter with respect to it.
+    */
```
done

mzumsande commented on src/addrman.cpp:
```diff
@@ -592,6 +599,7 @@ bool AddrManImpl::AddSingle(const CAddress& addr, const CNetAddr& source, std::c
         pinfo = Create(addr, source, &nId);
         pinfo->nTime = std::max(NodeSeconds{0s}, pinfo->nTime - time_penalty);
         nNew++;
+        m_map_counts[pinfo->GetNetwork()].first++;
```
Yes, I wanted to keep these together. I will add a commit that moves both of these lines into `Create`.

mzumsande commented on src/addrman.cpp:
```diff
@@ -962,6 +970,28 @@ std::optional<AddressPosition> AddrManImpl::FindAddressEntry_(const CAddress& ad
+        } else {
+            return *in_new ? entry.first : entry.second;
+        }
+    }
```
Yes, these need to be identical, but I actually changed it to use the old `vRandom.size()` used in `size()` so that isn't a change from the status quo when the two `size()` functions are merged as you suggest.

mzumsande commented on src/addrman.cpp:
```diff
@@ -962,6 +970,28 @@ std::optional<AddressPosition> AddrManImpl::FindAddressEntry_(const CAddress& ad
     }
 }
 
+size_t AddrManImpl::size_(std::optional<Network> net, std::optional<bool> in_new) const
```
Yes, I had used lower case for consistency. I now changed it to upper case, and added a commit that removes the old `size()` to always use the new `Size()` as suggested by Amiti. I made that an extra commit because it involved renaming a lot of existing call sites.

mzumsande commented:
Thanks a lot for the reviews! I think I have addressed all comments and implemented most of the suggestions - let me know if I missed something!

> another style related suggestion: what do you think about consolidating the two `size()` functions? adding default params to the newly introduced function means you can remove the old one & none* of the call sites need to change. I tried it out [here](https://github.com/amitiuttarwar/bitcoin/commit/ab188a62bd18668f5a9174db4dadfc398bc02b76).
> [*] none of the bitcoind callers. one test call site failed because of the shift from `vRandom.size` to `nNew + nTried`. the test is checking a corrupted addrman, and I don't think the size is testing anything meaningful there, so I removed that check.

I did that, and combined it with @vasild's suggestion to use upper case notation `Size()` as per the developer notes. This involved a lot of renaming of existing call sites, so I did it in an extra commit at the end. Since I switched back to use `vRandom.size` to keep existing behavior, the corrupted AddrMan test doesn't need to be changed anymore.

> Or ditch nNew and sum m_map_count[].first whenever we would read nNew, via a new method like size_t CountNew() const? Performance considerations?

> this one I'm more open to. martin pointed out some difficulties in that approach [here](https://github.com/bitcoin/bitcoin/pull/26847#discussion_r1065041916), and I personally think its nbd either way.

Yes, since it's part of the serialization we can't completely do without it. Of course, we could decide to just use `nNew` and `nTried` locally in `Unserialize()`, and rebuild it from the sum from `m_network_counts` everywhere else. I would suspect that there is no major difference performance-wise because the maximum number of networks is very small - I could try to benchmark it.

naumenkogs commented on src/addrman.cpp:
```diff
@@ -962,6 +970,28 @@ std::optional<AddressPosition> AddrManImpl::FindAddressEntry_(const CAddress& ad
+        }
+    }
+    if (auto it = m_network_counts.find(*net); it != m_network_counts.end()) {
+        auto entry = it->second;
```
nit: I'd rename this into `network_count`. `entry` made me think it's about one particular addr

mzumsande replied:
done

naumenkogs commented on src/addrman.cpp:
```diff
@@ -1065,6 +1098,17 @@ int AddrManImpl::CheckAddrman() const
+        return -20;
+    }
+    for (const auto& [net, count] : m_network_counts) {
+        if (local_counts[net].n_new != count.n_new || local_counts[net].n_tried != count.n_tried) {
```
`local_count[net]` could be called on the non-existent key, but that's fine and will return 0, right?

naumenkogs replied:
perhaps a test for this case would be great...

naumenkogs replied:
Oh, I should have tried to break it and see whether there is a such test. Explicit is better of course.

mzumsande replied:
For this to happen, we need to have an address in AddrMan for a network at one point in time, which later gets removed, so that we have an entry in `m_network_counts` that has only zeroes, but no entry in `local_counts`.
This happens in the `remove_invalid` subtest which tampers with the serialized stream, so that an address gets added and then removed during Deserialization - if you break `CheckAddrman()` by adding an assert at this spot, this subtest will fail.

I think it could also happen theoretically if the last remaining address from network A is being kicked out by an address from network B due to a collision in the new tables. But this doesn't seem trivial to write a unit test for, since addresses from different networks are typically from different groups, so I'm not sure if it's easily possible to write an explicit test for this.

naumenkogs commented:
utACK 80f39c9

stratospher commented:
ACK 80f39c9

vasild commented on src/addrman.h:
```diff
@@ -99,8 +99,14 @@ class AddrMan
+    * @param[in] in_new           Select addresses only from one table (true = new, false = tried, nullopt = both)
+    * @return                     Number of unique addresses that match specified options.
+    */
+    size_t Size(std::optional<Network> net = {}, std::optional<bool> in_new = {}) const;
```
nit, if you re-touch: usually `std::nullopt` is used instead of `{}` and I think it is more readable, e.g.:

https://github.com/bitcoin/bitcoin/blob/79e18ebc81f112f989ff50fb6f5a6a17db9b6758/src/streams.h#L236

mzumsande replied:
will do if I need to re-touch for another reason.

mzumsande replied:
done in #27015

vasild reviewed (approved):
ACK 80f39c99ef2d30e3e2d8dbc068d25cf92aa32344

achow101 commented:
ACK 80f39c99ef2d30e3e2d8dbc068d25cf92aa32344

mzumsande commented:
I just noticed that this can lead to intermittent failures of the `load_addrman_corrupted` unit test, while checking the size in
https://github.com/bitcoin/bitcoin/blob/ba3d32715f985314eb1fdb006bfc4127f8d516a7/src/test/addrman_tests.cpp#L951 the internal checks fail (that weren't executed before this PR). I think it might be best to just remove this line from the test, since this is a situation where we purposefully load a corrupted `peers.dat` - seems a good thing that the Addrman Check can fail then... I will open a PR tomorrow.
